### PR TITLE
Move Folia region scheduler API to Paper for easy compat

### PR DESCRIPTION
**Pull request by kennytv**
Docs might need slight adjustment first to tell people they should only really use the schedulers when also targetting Folia, otherwise this keeps almost all logic of the actual Folia schedulers intact (excluding RegionScheduler) to stay consistent

### CODE REVIEW

**Comment by olijeffers0n** [comment]
Handy, I might work on a docs page at some point to just outline how you should make plugins interoperable with Paper/Folia.

A lot of the ways people check for paper is with a `Class.forName` call, and this will break all of those plugins. Can we have a `Server/Bukkit#isFolia` or some method along those lines if we are going to blur the line between the two by pulling this patch into Paper anyways? Would alleviate a lot of the headache when this gets merged is all.

File: patches/server/0976-Folia-scheduler-API.patch [hunk 1]
```diff
@@ -0,0 +1,1269 @@
...
++import java.util.concurrent.ThreadFactory;
++import java.util.concurrent.ThreadPoolExecutor;
++import java.util.concurrent.TimeUnit;
++import java.util.concurrent.atomic.AtomicInteger;
++import java.util.function.Consumer;
++import java.util.logging.Level;
++
++public final class FoliaAsyncScheduler implements AsyncScheduler {
++
++    private static final Logger LOGGER = LogUtils.getLogger();
```
Comment (Machine-Maker) [inline patches/server/0976-Folia-scheduler-API.patch]: I think we prefer the getClassLogger as it doesn't have a huge prefix.

**Comment by stonar96** [comment]
Can we also add `Server#isOwnedByCurrentRegion(...)`?

**Comment by kennytv** [comment]
The owned region methods have been added, minus the global region tick thread check. An isFolia method would be a bit silly, since these methods are pulled exactly so that you *don't* need to have any unnecessary Folia checks. Plugins can/should still check this on their own if they need additional Folia methods, whether through the server name, by checking whether methods/classes they actually need exist, or checking for `io.papermc.paper.threadedregions.RegionizedServer`

**Comment by stonar96** [comment]
> The owned region methods have been added, minus the global region tick thread check. An isFolia method would be a bit silly, since these methods are pulled exactly so that you _don't_ need to have any unnecessary Folia checks. Plugins can/should still check this on their own if they need additional Folia methods, whether through the server name, by checking whether methods/classes they actually need exist, or checking for `io.papermc.paper.threadedregions.RegionizedServer`

First of all thanks for adding the `isOwnedByCurrentRegion` methods. Regarding `isFolia` I only partially agree, since (I think) I have a valid usecase: Let's assume you have a `BukkitRunnable` in your Paper plugin that does something for each player on the main thread. Now you would have to schedule a task for each player to make it compatible with Folia (afaik). In this case I'd prefer to be able to check if `isFolia` and if `false`, still use a single `BukkitRunnable`.

**Comment by imDaniX** [comment]
I'd say `isFolia` should just be a part of [PaperLib](https://github.com/PaperMC/PaperLib), not Paper itself.

**Comment by stonar96** [comment]
Agree, didn't think about that.

**Comment by olijeffers0n** [comment]
Yeah, I mentioned it because of cat's comment on a PaperLib PR from the other day:
> PaperLib is concerned about paper vs spigot, idk if there is any intent to expand this onto paper vs spigot vs folia

So it seemed like maybe this was the more appropriate place

**Comment by Malfrador** [comment]
I don't really want to use PaperLib for a Folia check.  I don't care about Spigot support, and needing to use a lib with many spigot-related features would be unnecessary. 
There certainly are situations in which I would like to check if the server is running Folia or not, even if the Scheduler methods are in Paper. For example I might want to run different logic that only makes sense if there are multiple Folia regions. 

Checking for the server implementation name doesn't work with forks, and needing to check for the existence of RegionizedServer is just an awkward solution. So I really would like an isFolia method as well.
Folia and Paper being worked on by the same people is a different situation than Spigot/Paper, so there shouldn't be any workarounds like checking for the existence of a specific class needed.

**Comment by electronicboy** [comment]
I would much rather have a feature flag system as opposed to some asinine "isX" set of methods, having methods to check what platform you're on would generally suggest a flaw in how the API is working or in how you're using it, you'd still need to handle aspects of those methods existing or not unless you'reonly catering towards the latest version of paper, etc

**Comment by F3F5** [comment]
i love this pull

**Review by ghost — APPROVED** [review]
like

**Comment by kennytv** [comment]
Feature flags do sound better, but out of scope for this pr, so here I go

**Pull request merged**